apps/astarte_appengine_api/lib/astarte_appengine_api_web/router.ex

```diff
@@ -19,32 +19,70 @@ defmodule Astarte.AppEngine.APIWeb.Router do
   use Astarte.AppEngine.APIWeb, :router
   alias Astarte.AppEngine.APIWeb.Plug.LogRealm
 
-  pipeline :api do
+  pipeline :auth_api do
     plug :accepts, ["json"]
     plug LogRealm
     plug Astarte.AppEngine.APIWeb.Plug.AuthorizePath
   end
 
+  pipeline :device_api do
+    plug :auth_api
+    plug Astarte.AppEngine.APIWeb.Plug.LogDeviceId
+  end
+
+  pipeline :device_alias_api do
+    plug :auth_api
+    plug Astarte.AppEngine.APIWeb.Plug.LogDeviceAlias
+  end
+
+  pipeline :device_group_api do
+    plug :auth_api
+    plug Astarte.AppEngine.APIWeb.Plug.LogGroupName
+  end
+
+  pipeline :interface_api do
+    plug Astarte.AppEngine.APIWeb.Plug.JoinPath
+    plug Astarte.AppEngine.APIWeb.Plug.LogInterface
+    plug Astarte.AppEngine.APIWeb.Plug.LogPath
+  end
+
+  pipeline :interface_values_api do
+    plug :device_api
+    plug :interface_api
+  end
```

Comment:
I think the rationalization of the APIs looks promising but I will leave a few considerations.
I am not convinced about:
- defining pipelines of pipelines of plugs
- having several `scope`s at the same level just to use different pipelines/plugs on them

Indeed, the router now defines different scopes, all at the same level.
Each scope uses a pipeline of plugs.
Each pipeline of plugs can consist of other pipeline of plugs, which in turn can consist of other pipelines of plugs, and so on.

My doubt is that it might be clearer to instead have nested `scope`s. This way:
1. you don't have to redefine the route of each scope from scratch but rely on the fact that the structure of the routes is already effectively nested
2. in the outer scope you can insert the more generic plugs about the realm, but you don't have to insert them again for each nested scope.
3. by nesting scopes instead of pipelines you also have a visual feedback on the structure of the routes. At the same time, it's easier to spot which plugs are called for each route without having to follow through the definition of pipelines of pipelines of pipelines of plugs.

Reply:
Every controller in Astarte has a specific scope of plugs. Some controllers have similar plugs but with one or two differences (e.g., controllers for interfaces). I try to avoid duplicate code by organizing plugs into groups and using pipelines, such as base pipelines, for other pipelines. (The base pipeline is auth_api).

Reply:
I think @davidebriani is suggesting that it would be clearer to have nested scopes, possibly adding plugs at each nesting level, rather than using pipelines as plugs inside other pipelines. I do not have a strong opinion on this, however the nested scopes approach could be more clear as it allows to segregate the different sections of the API without adding new names to keep in mind (i.e. pipeline names)